elastalert/test_rule.py

```diff
@@ -23,63 +36,244 @@ def print_terms(terms, parent):
 
-        if args.schema_only:
+    # Separate _source and _id, convert timestamps
+    resp = [{'_source': doc, '_id': doc['_id']} for doc in docs]
+    [doc['_source'].pop('_id') for doc in resp]
+    replace_hits_ts(resp, rule)
+    return resp
+
+
+def mock_terms(rule, start, end, index, key, qk=None, size=None):
+    """ Mocks the effects of get_hits_terms using global data instead of Elasticsearch. """
+    if key.endswith('.raw'):
+        key = key[:-4]
+    buckets = {}
+    for doc in data:
+        if key not in doc:
             continue
+        if start <= ts_to_dt(doc[rule['timestamp_field']]) < end:
+            if qk is None or doc[rule['query_key']] == qk:
+                buckets.setdefault(doc[key], 0)
+                buckets[doc[key]] += 1
+    counts = buckets.items()
+    counts.sort(key=lambda x: x[1], reverse=True)
+    if size:
+        counts = counts[:size]
+    buckets = [{'key': value, 'doc_count': count} for value, count in counts]
+    return {end: buckets}
+
+
+def mock_elastalert(elastalert):
+    """ Replaces elastalert's get_hits functions with mocks. """
+    elastalert.get_hits_count = mock_count
+    elastalert.get_hits_terms = mock_terms
+    elastalert.get_hits = mock_hits
+    elastalert.new_elasticsearch = mock.Mock()
+
+
+def run_elastalert(args):
+    """ Creates an ElastAlert instance and run's over for a specific rule using either real or mock data. """
+    # Mock configuration. Nothing here is used except run_every
+    conf = {'rules_folder': 'rules',
+            'run_every': datetime.timedelta(minutes=5),
+            'buffer_time': datetime.timedelta(minutes=10),
+            'alert_time_limit': datetime.timedelta(hours=24),
+            'es_host': 'es',
+            'es_port': 14900,
+            'writeback_index': 'wb',
+            'max_query_size': 100000,
+            'old_query_limit': datetime.timedelta(weeks=1),
+            'disable_rules_on_error': False}
+
+    # Load and instantiate rule
+    with open(args.file) as fh:
+        rule = yaml.load(fh)
+    load_options(rule)
+    load_modules(rule)
+    conf['rules'] = [rule]
+
+    # If using mock data, make sure it's sorted and find appropriate time range
+    timestamp_field = rule.get('timestamp_field', '@timestamp')
+    if args.json:
+        global data
+        if not data:
+            return
+        try:
+            data.sort(key=lambda x: x[timestamp_field])
+            starttime = ts_to_dt(data[0][timestamp_field])
+            endtime = data[-1][timestamp_field]
+            endtime = ts_to_dt(endtime) + datetime.timedelta(seconds=1)
+        except KeyError as e:
+            print("All documents must have a timestamp and _id", e)
+            return
+
+        # Create mock _id for documents if it's missing
+        def get_id():
+            return ''.join([random.choice(string.letters) for i in range(16)])
+
+        [doc.update({'_id': doc.get('_id', get_id())}) for doc in data]
```

Comment:
As before, please split this out instead of putting a mutator in the body of a list comprehension.